internal/gengapic/paging.go

```diff
@@ -67,8 +57,33 @@ func (g *generator) iterTypeOf(elemField *descriptor.FieldDescriptorProto) (*ite
 			typ = parent
 		}
 
-		pt.elemTypeName = fmt.Sprintf("*%s.%s", imp.Name, typeName)
-		pt.iterTypeName = typeName + "Iterator"
+		eMsg, ok := eType.(*descriptor.DescriptorProto)
+		if !ok {
+			return nil, errors.E(nil, "cannot find message type %q, malformed descriptor", eType)
+		}
+		if eMsg.GetOptions().GetMapEntry() {
+			var valueField *descriptor.FieldDescriptorProto
+			for _, f := range eMsg.GetField() {
+				if f.GetName() == "value" {
+					valueField = f
+					break
+				}
+			}
+			if valueField == nil {
+				return nil, errors.E(nil, "unusual map entry message: %q", eMsg)
+			}
+
+			if valueField.GetType() == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
+				pt.mapValueTypeName = valueField.GetTypeName()
+			} else {
+				pt.mapValueTypeName = pbinfo.GoTypeForPrim[valueField.GetType()]
+			}
+			pt.elemTypeName = fmt.Sprintf("String%sPair", upperFirst(pt.mapValueTypeName))
```

Comment:
I feel like encoding the types into the struct name is a little unnecessary. Maybe just `%Entry`? like `AcceleratorTypesScopedListEntry`. `Pair` is fine too, or `Item`. I just don't think the `String` is necessary.

Reply:
I am leaning towards `Pair` more than `Entry` to make it more clear that this struct is subtly distinct from the MapEntry message generated by protoc since the types are coalesced. I'll remove `String`.

Reply:
> the MapEntry message generated by protoc

That `MapEntry` message is an internal detail of the descriptors presented by protoc to a plugin. It doesn't make it into the generated Go code from protobuf-go, so there is nothing to differentiate it from in the generated output. But `Pair` is fine.

Reply:
>  It doesn't make it into the generated Go code from protobuf-go
Correct, and I wanted to make it clear that this is not that. I'll stay with `Pair`.